### DISCUSSION TITLE
get detailed role object in module's preRegisterCheck function

### DISCUSSION THREAD

**Discussion by peauc** [directus/directus#16720] [Archive]
Hello everyone, I have an issue with restricting permissions on a custom module in Directus.
I want to restrict one module's access for a certain kind of users. 
I have defined a few fields in the `directus_roles` table as metadata that I cannot seem to access in the [function `preRegisterCheck` of my module](https://docs.directus.io/extensions/modules.html#extension-entrypoint) . 
Is there a way to accomplish that ?

How could I restrict the module's access otherwise ? I'm tempted to fork the project to change the [hydration of the currentUser object](https://github.com/directus/directus/blob/7bf90efa6243f60fa4128a82540f2995df2ce9f5/app/src/stores/user.ts#L37) to add my metadata there.

🙏 Thank you for taking the time to read my question and for this great OSS.

**Comment by rijkvanzanten** [comment]
As of right now, the preRegisterCheck function gets called with the current user, and that user's permissions:

https://github.com/directus/directus/blob/0dda79d53184610d622890f0f473d6f1353779a1/app/src/modules/index.ts#L33

There's some light information about the role available on the user, but it's hardcoded to a subset of fields required for app operation:

https://github.com/directus/directus/blob/0dda79d53184610d622890f0f473d6f1353779a1/app/src/stores/user.ts#L48-L51

That being said, seeing that the stores are already hydrated by the time this is called, I think you should be able to use the rolesStore to get the custom information you're looking for

**Reply by peauc** [reply]
I have tried accessing the stores but encounter an error. My inexperienced opinion is that the store objects exposed in the `extension-sdk` are not yet available. Am i doing something wrong ?
Same error for the useApi btw.
![CleanShot 2022-12-05 at 16 19 14](https://user-images.githubusercontent.com/15132860/205673975-b594a894-e4d5-4b83-b69f-5c6be5130bd5.png)
![CleanShot 2022-12-05 at 16 22 19](https://user-images.githubusercontent.com/15132860/205674663-608f6eaa-5ec1-4ffc-9b70-c1571af606f8.png)

**Reply by flevi29** [reply]
I can confirm this is still the case for 9.22.4.
```
{
  "fileName": "http://localhost:3002/extensions/sources/index.js",
  "lineNumber": 5,
  "columnNumber": 11,
  "message": "[useStores]: The stores could not be found."
}
```
The next best shot seems to be to write custom logic on the component that displays a big message like "You don't have permission to use this module..." sort of thing for now.

**Reply by flevi29** [reply]
Well then, I am not even sure how to get working what I just suggested. The following code within my module.vue
```typescript
import { useApi, useStores } from "@directus/extensions-sdk";

const { useUserStore, useFieldsStore } = useStores();
const userStore = useUserStore();
const { id } = userStore.currentUser.role;
console.log(id);
// Didn't see any `rolesStore` in the code, so I assume this is another way of doing it
const fieldsStore = useFieldsStore();
console.log(fieldsStore.getField("directus_roles", id));
```
prints `b06dcce5-9555-4d0f-9c54-b8c0958eb698` and `null`.
Am I querying with the wrong permissions or what? How would that even make any sense, I am doing this as the main admin role. I guess I'll try with `useApi` next.

**Reply by flevi29** [reply]
No permission with `useApi` either. I guess there's really no straight forward easy way to do this.

**Reply by flevi29** [reply]
I had to make a custom endpoint extension, which only does one thing: verify the role of the current user and based on that respond with a boolean. Now my suggestion works.

**Comment by peauc** [comment]
I don't think what I'm trying to accomplish is possible, I can't seem to be able to access the directus api/sdk from the registerCheck method.
I'm trying a hack-y solution with collection permissions to check the access.

Otherwise next step is to fork the project and change the hydration myself :(. 

On a side note, would it be possible to pass the role name `currentUser` ?? It would solve a use case for us.

**Comment by flevi29** [comment]
It would be nice if we could make further checks in `preRegisterCheck` of the users role. I detailed my [workaround](https://github.com/directus/directus/discussions/16720#discussioncomment-4860146), but it'd be better and more appropriate if that hook would handle this I suppose. Only problem is that the stores and api is not accessible from that function, which might be a bug, since @rijkvanzanten said it should be possible to use them there.
> That being said, seeing that the stores are already hydrated by the time this is called, I think you should be able to use the rolesStore to get the custom information you're looking for

**Comment by peauc** [comment]
@flevi29 i have a workaround, I created a collection named "full_access". Then I just have to check if user has create access to this collection in the pre-register check. If user has only create permissions on a collection it doesn't show up in the collection view.
This is less than satisfactory but ...

**Reply by flevi29** [reply]
Thanks for the details, I'll try submitting a reproduction and issue about `preRegisterCheck` stores error later, and if nothing can be done about it I'll apply your fix.

**Reply by flevi29** [reply]
I ended up using a hook instead which works better. Basically you conditionally edit the payload on the fly, specifically the `module_bar` property so that it has your module set to disabled.
```typescript
import { defineHook } from "@directus/extensions-sdk";

export default defineHook(
  async (
    { filter },
    {
      services,
      getSchema,
      exceptions: { InvalidPayloadException },
    },
  ) => {
    const { ItemsService } = services;
    const schema = await getSchema();
    const usersService = new ItemsService("directus_users", { schema });

    // disables Harvest Manager module for those who have no access to it
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    filter("settings.read", async (items: any, _, { accountability }) => {
      if (
        accountability !== null &&
        items[0]?.constructor === Object &&
        Object.hasOwn(items[0], "module_bar")
      ) {
        const [{ harvest_manager_access }] = await usersService.readByQuery({
          fields: ["harvest_manager_access"],
          filter: { id: { _eq: accountability.user } },
        });
        if (typeof harvest_manager_access !== "boolean") {
          throw new InvalidPayloadException(
            `"harvest_manager_access" boolean could not be found on user (${harvest_manager_access})`,
          );
        }
        if (!harvest_manager_access) {
          for (const module of items[0].module_bar) {
            if (module.id === "harvest_manager" && module.enabled) {
              module.enabled = false;
              break;
            }
          }
        }
      }
    });
  },
);
```

Only downside so far is that this is very custom and may break in the future, if the Directus team made `preRegisterCheck` a little more useful this wouldn't be needed, but as long as it works.

**Reply by flevi29** [reply]
Another downside seems to be that they can still access the module by just typing it's URL path into the browser. But if you also restrict every endpoint call your module uses, then it shouldn't be such a big deal.
